contracts/chain-adapters/ForwarderBase.sol

```diff
@@ -0,0 +1,240 @@
+        if (nextLayerContracts.adapter == address(0)) revert UninitializedRoute();
+        bool success;
+        // Case 1: We are immediately before the target spoke pool, so we send the tokens to the spoke pool
+        // and do NOT follow it up with a message.
+        if (target == nextLayerContracts.spokePool) {
+            (success, ) = nextLayerContracts.adapter.delegatecall(
+                abi.encodeCall(AdapterInterface.relayTokens, (baseToken, destinationChainToken, amount, target))
+            );
+            if (!success) revert RelayTokensFailed(baseToken);
+        } else {
+            // Case 2: We are not immediately before the target spoke pool, so we send the tokens to the next
+            // forwarder and accompany it with a message containing information about its intended destination.
+
+            if (nextLayerContracts.forwarder == address(0)) revert MalformedRoute();
+            // Send tokens to the forwarder.
+            (success, ) = nextLayerContracts.adapter.delegatecall(
+                abi.encodeCall(
+                    AdapterInterface.relayTokens,
+                    (baseToken, destinationChainToken, amount, nextLayerContracts.forwarder)
+                )
+            );
+            if (!success) revert RelayTokensFailed(baseToken);
+
+            // Send a follow-up message to the forwarder which tells it to continue bridging.
+            bytes memory message = abi.encodeCall(
+                AdapterInterface.relayTokens,
+                (baseToken, destinationChainToken, amount, target)
+            );
+            (success, ) = nextLayerContracts.adapter.delegatecall(
+                abi.encodeCall(AdapterInterface.relayMessage, (nextLayerContracts.forwarder, message))
+            );
+            if (!success) revert RelayMessageFailed();
+        }
+        emit TokensForwarded(target, baseToken, amount);
+    }
+
+    // Function to be overridden in order to authenticate that messages sent to this contract originated
+    // from the expected account.
+    function _requireAdminSender() internal virtual;
+
+    // We also want to restrict who can upgrade this contract. The same admin that can relay messages through this
+    // contract can upgrade this contract.
+    function _authorizeUpgrade(address) internal virtual override onlyAdmin {}
+
+    function _setCrossDomainAdmin(address _newCrossDomainAdmin) internal {
+        if (_newCrossDomainAdmin == address(0)) revert InvalidCrossDomainAdmin();
+        crossDomainAdmin = _newCrossDomainAdmin;
+        emit SetXDomainAdmin(_newCrossDomainAdmin);
+    }
+
+    // Reserve storage slots for future versions of this base contract to add state variables without
+    // affecting the storage layout of child contracts. Decrement the size of __gap whenever state variables
+    // are added. This is at bottom of contract to make sure it's always at the end of storage.
+    uint256[1000] private __gap;
```

Comment:
Should this be relocated above the function and event definitions? It seems more likely that any subsequent additions would be added before them and might overlook the existence of this gap.

Reply:
I think you raise a good point but for consistency let's leave it at bottom of contract. Any developer familiar with our SpokePool will know to look for it down here

Reply:
Ah sorry; I was mixed up and thought it was the opposite. Agree 100%: